### DISCUSSION TITLE
Easier type annotations for root_validator values argument in v1?

### DISCUSSION THREAD

**Discussion by kaysond** [pydantic/pydantic#6227] [Question] [answered]
I'm writing a `root_validator` for a model with some shallow-ly nested types. I need to annotate the `values` type, but since it's a `dict`, I end up with a giant mess, and I have to maintain a second copy of the types from the model definition, which is dangerous.

Is there a better way of doing this? I see that v2 changes this to pass it in as the model type itself, but what about v1?

Thanks!

```python
class MyModel(BaseModel):
    param1: StrictStr
    param2: list[StrictStr]
    param3: dict[
        str, list[StrictStr] | list[StrictInt] | list[StrictFloat] | AnotherModel
    ]
    param4: list[StrictStr]

    @root_validator
    @classmethod
    def cross_check(
        cls,
        values: dict[
            str,
            StrictStr
            | list[StrictStr]
            | dict[
                str,
                list[StrictStr]
                | list[StrictInt]
                | list[StrictFloat]
                | AnotherModel,
            ],
        ],
    ):
        pass

**Comment by adriangb** [comment, marked as answer]
Unfortunately, I don't think there is a good solution to this in V1, at least not one I've found.

**Reply by kaysond** [reply]
Is there an ETA for v2?

**Reply by adriangb** [reply]
The next couple of months. But we don't expect much will change compared to the beta release so please evaluate it, it may be ready for your use case!
